sklearn/datasets/openml.py

```diff
@@ -77,20 +77,47 @@ def _open_openml_url(openml_path, data_home):
 
     Returns
     -------
-    result : stream
-        A stream to the OpenML resource
+    result : bytes
+        Byte content of resource
     """
     def is_gzip(_fsrc):
         return _fsrc.info().get('Content-Encoding', '') == 'gzip'
 
     req = Request(_OPENML_PREFIX + openml_path)
     req.add_header('Accept-encoding', 'gzip')
 
+    def _md5_validated_bytes(bytes_content, md5_checksum):
+        """
+        Consume binary stream to validate checksum,
+        return a new stream with same content
+
+        Parameters
+        ----------
+        bytes_content : bytes
+
+        md5_checksum: str
+            Expected md5 checksum of bytes
+
+        Returns
+        -------
+        bytes
+        """
+        actual_md5_checksum = hashlib.md5(bytes_content).hexdigest()
```

Comment:
I'm wondering if we should checksum in in chunks. Could you please take a medium sized OpenML dataset (e.g. MNIST https://www.openml.org/d/554 ) and see computing the checksum with the above way is still reasonable with respect to memory usage (see `memory_profiler`).

Reply:
It does not make any difference.

```bash
(venv) shashank@precision-7510:~/scikit-learn/scikit-learn$ cat md5_mem_test.py 
import sys
from hashlib import md5

import requests

from memory_profiler import profile
from sklearn.datasets import fetch_openml


@profile(precision=3)
def hash_block():
    stream = requests.get("https://www.openml.org/data/download/52667/mnist_784.arff").content
    return md5(stream).hexdigest()


@profile(precision=3)
def hash_chunked(chunk_size=512):
    md5sum = md5()
    stream = requests.get("https://www.openml.org/data/download/52667/mnist_784.arff").content
    for i in range(0, len(stream), chunk_size):
        md5sum.update(stream[i : i + chunk_size])
    return md5sum.hexdigest()


if __name__ == "__main__":
    if len(sys.argv) > 1:
        print(hash_chunked())
    else:
        print(hash_block())

(venv) shashank@precision-7510:~/scikit-learn/scikit-learn$ python -m memory_profiler md5_mem_test.py
Filename: md5_mem_test.py

Line #    Mem usage    Increment   Line Contents
================================================
    10   65.137 MiB   65.137 MiB   @profile(precision=3)
    11                             def hash_block():
    12  311.012 MiB  245.875 MiB       stream = requests.get("https://www.openml.org/data/download/52667/mnist_784.arff").content
    13  311.012 MiB    0.000 MiB       return md5(stream).hexdigest()


0298d579eb1b86163de7723944c7e495
(venv) shashank@precision-7510:~/scikit-learn/scikit-learn$ python -m memory_profiler md5_mem_test.py chunked
Filename: md5_mem_test.py

Line #    Mem usage    Increment   Line Contents
================================================
    16   65.645 MiB   65.645 MiB   @profile(precision=3)
    17                             def hash_chunked(chunk_size=512):
    18   65.645 MiB    0.000 MiB       md5sum = md5()
    19  310.742 MiB  245.098 MiB       stream = requests.get("https://www.openml.org/data/download/52667/mnist_784.arff").content
    20  310.742 MiB    0.000 MiB       for i in range(0, len(stream), chunk_size):
    21  310.742 MiB    0.000 MiB           md5sum.update(stream[i : i + chunk_size])
    22  310.742 MiB    0.000 MiB       return md5sum.hexdigest()


0298d579eb1b86163de7723944c7e495

```

I guess it makes sense,  from the wiki [page](https://en.wikipedia.org/wiki/MD5):

>  The main algorithm then uses each 512-bit message block in turn to modify the state

The only optimization that can be done here imo is that if the md5checksum flag is enabled (default), we could download the arff file in chunks and keep updating md5, which would save us some time. It might add some code complexity though (we will have to manage the chunking with [urllib](https://docs.python.org/3/library/urllib.request.html)), do the maintainers think it is worth it?

Reply:
Currently the data is traversed twice. First when the stream is read into memory. Second when `hashlib.md5` goes through the data.

Would the goal of this optimization make it so we read the data twice?

In your testing do you find that the checksum increases the time it loads the data?

Reply:
Added the optimization (it now reads the stream only once, and keeps updating the md5 while reading). 

In terms of memory, it will not add any significant overhead as I mentioned, since md5 uses a state-machine.

In terms of time / cpu, here is a quick test result for a medium size dataset:

Existing openml (master branch)
```
$ python -m timeit -c "from sklearn.datasets import fetch_openml; fetch_openml(data_id=554, data_home=None)"
1 loop, best of 5: 12.4 sec per loop
```

This PR:
```
$ python -m timeit -c "from sklearn.datasets import fetch_openml; fetch_openml(data_id=554, data_home=None, verify_checksum=False)"
1 loop, best of 5: 12.3 sec per loop

$ python -m timeit -c "from sklearn.datasets import fetch_openml; fetch_openml(data_id=554, data_home=None, verify_checksum=True)"
1 loop, best of 5: 12.6 sec per loop
```

This is single iteration but it looks like it is not making it significantly slow

Reply:
What was the time before the optimization? (https://github.com/scikit-learn/scikit-learn/pull/14800/commits/831d78b57bbc9f5f87175ad502d9aa37710d6193)

Reply:
https://github.com/scikit-learn/scikit-learn/commit/831d78b57bbc9f5f87175ad502d9aa37710d6193 was slightly faster for big datasets (12.1 sec vs 12.3 sec) but slower for normal datasets. (1.49 msec vs 1.39msec). This is only some manual calls with timeit on the shell. Haven't done exhaustive testing. Should I? I am not sure if there is a significant difference

Reply:
I do not think this actually is better than before. The way to get a benefit from chunking is to verify the data the same time it is being encoded by the arrf parser.

The chunking this PR is doing still end up reading the data twice. Once, for checking, and another when the data is being parsed by the arrf parser.

Reply:
Data was already being read twice, once [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/datasets/openml.py#L466) and once again [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/externals/_arff.py#L793). 

This PR maintains the number of times the stream is read. 

To make this validation occur during decoding the arff would:
1. Be wasteful if done each time decode is called (currently validation is done only when data is downloaded from internet)
2. Would involve adding check in the arff decode logic which is in the externals module.
Would we prefer to modify this?

Reply:
I would not prefer not changing the arrf module. I suspect the parsing of the arrf file needs the whole thing in memory. I am happy with previous non chucked version.

Reply:
https://github.com/scikit-learn/scikit-learn/commit/831d78b57bbc9f5f87175ad502d9aa37710d6193 would make the stream be read 3 times, once at download (non chunked), then to validate (all at one go) and then to decode (arff). 

In my testing the second validation overhead is insignificant in terms of memory or cpu, so I would not worry too much about it. If you also agree then I will push a revert commit

Reply:
Okay, lets keep the current version.